File: Chapter 03/Python/section_3_4_order_processor_refactor/after/order_service_components.py
"""
SRP SOLUTION: Service Isolation.
* ARCHITECTURE NOTE: Each class here represents a specific domain boundary. 
The validator only knows how to check data, the payment service only 
knows how to talk to a gateway, etc. 
* This isolation makes each component independently testable and reusable 
across different parts of the application.
"""

class OrderValidator:
    def validate(self, order):
        print("  [Validate] Validating order...")
        if not order.items or order.total <= 0:
            raise ValueError("Order is invalid.")

class PaymentService:
    def process_payment(self, order):
        print(f"  [Payment] Processing payment for ${order.total:.2f}...")
        return True

class InventoryManager:
    def update_inventory(self, order):
        print("  [Inventory] Updating inventory...")

class NotificationService:
    def send_confirmation_email(self, order):
        print(f"  [Notify] Sending confirmation email to {order.customer_email}...")

File: Chapter 03/C#/Section_3_4_OrderProcessor_Refactor/After/OrderServiceComponents.cs
using System;

namespace Chapter03.OrderProcessorRefactor.After;

// --- Step 1: The Individual Service Classes ---

/// <summary>
/// SRP SOLUTION: Domain Logic Isolation.
/// 
/// ARCHITECTURE NOTE: This class has one job: validation. It doesn't know 
/// about databases or payment gateways. This makes it incredibly easy to 
/// unit test our business rules without needing mock APIs or databases.
/// </summary>
public class OrderValidator
{
    public void Validate(Order order)
    {
        Console.WriteLine("  [Validate] Validating order...");
        if (order.Items.Count == 0 || order.Total <= 0)
        {
            throw new InvalidOperationException("Order is invalid.");
        }
    }
}

/// <summary>
/// SRP SOLUTION: External API Isolation.
/// 
/// ARCHITECTURE NOTE: If Stripe or PayPal changes their API, this is the 
/// ONLY file that needs to be updated. The rest of the checkout process 
/// remains completely untouched and safe.
/// </summary>
public class PaymentService
{
    public bool ProcessPayment(Order order)
    {
        Console.WriteLine($"  [Payment] Processing payment for ${order.Total}...");
        // Real payment gateway logic would go here
        return true;
    }
}

/// <summary>
/// SRP SOLUTION: Infrastructure Isolation.
/// 
/// ARCHITECTURE NOTE: If we migrate from SQL Server to PostgreSQL, the 
/// billing team and sales team won't even notice, because only this 
/// inventory manager class needs to change.
/// </summary>
public class InventoryManager
{
    public void UpdateInventory(Order order)
    {
        Console.WriteLine("  [Inventory] Updating inventory...");
        // Real database logic to update stock would go here
    }
}

/// <summary>
/// SRP SOLUTION: Communications Isolation.
/// 
/// ARCHITECTURE NOTE: Email formatting and delivery logic lives purely here.
/// If an email template fails to render, it will no longer crash the entire 
/// payment transaction!
/// </summary>
public class NotificationService
{
    public void SendConfirmationEmail(Order order)
    {
        Console.WriteLine($"  [Notify] Sending confirmation email to {order.CustomerEmail}...");
        // Real email sending logic would go here
    }
}